debian/rcs2log.1
.\" $Debian: rcs2log.1,v 1.1 2001/06/14 07:24:39 epg Exp $
.\"
.TH RCS2LOG 1 "March 2001" "Debian Project" "Generates changelog entries"

.SH NAME
rcs2log \- generates a changelog from RCS files

.SH SYNOPSIS
.B rcs2log
[OPTION] ... [FILE ...]

.SH DESCRIPTION
.B rcslog 
generates a change log from RCS files, which can possibly be located
in a CVS repository, and outputs the new change log to standard
output.  If no FILE is specified, 
.B rcs2log
will use the files registered under the working directory.  The change
log prefix can be later copied to the existing ChangeLog file.

.SS OPTIONS

.TP 2
\fB\-c\fP CHANGELOG
Output a change log prefix to CHANGELOG (default ChangeLog).

.TP 2
\fB\-h\fP HOSTNAME  
Use HOSTNAME in change log entries (default current host).

.TP 2
\fB\-i\fP INDENT
Indent change log lines by INDENT spaces (default 8).

.TP 2
\fB\-l\fP LENGTH
Try to limit log lines to LENGTH characters (default 79).

.TP
.BI -R  
If no FILEs are given and RCS is used, recurse through working directory.

.TP
.BI -r 
OPTION  
Pass OPTION to subsidiary log command.

.TP 2
\fB\-t\fP TABWIDTH  
Tab stops are every TABWIDTH characters (default 8).

.TP 2
\fB\-u\fP "LOGIN\\tFULLNAME\\tMAILADDR"  
Assume LOGIN has FULLNAME and MAILADDR.

.TP
.BI -v  
Append RCS revision to file names in log lines.

.TP
.BI --help  
Output help.

.TP
.BI --version  
Output version number.

.SH AUTHOR
.B rcs2log
was written by Paul Eggert <eggert@twinsun.com>.

.SH COPYRIGHT
.B rcs2log
is free; anyone may redistribute copies of 
.B rcslog
to anyone under the terms stated in the GNU General Public License as
published by the Free Software Foundation; either version 2, or (at
your option) any later version.

.SH BUGS
There is a mailing list, bug-gnu-emacs@gnu.org on the internet, for
reporting bugs in rcs2log.  Please, if you find any bug, report it (if
you really are good-willing, a patch is appreciated).

Please note that this program is distributed in the hope that it will
be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.  In other terms, if
.B rcs2log
breaks, you get to keep both pieces.

.SH SEE ALSO
cvs(1),
cvsconfig(1)

.SH NOTE
This manual page was written by Andrea Glorioso <sama@aglorioso.com>
for the Debian GNU/Linux distribution because the original program
did not have a manual page.  


ChangeLog
2004-08-19 09:28  tmh

	* version_no.h, doc/build: Update version

2004-08-18 09:56  tmh

	* src/history.c: Remove history -X

2004-08-11 18:01  tmh

	* configure: break RH sserver/gsserver so every other OS still
	  works

2004-08-11 17:58  tmh

	* configure.in, redhat/cvsnt.spec: All the RH fixes now mean it
	  doesn't compile correctly on any other architecture.	Better to
	  break RH.

2004-08-11 17:30  tmh

	* redhat/cvsnt.spec.in: Over an hour for an OS that is <0.1% of
	  usage.  No comment.

2004-08-11 17:14  tmh

	* redhat/cvsnt.spec.in: FFS am I going to have to go through this
	  every time I want to release an RPM???

2004-08-11 17:01  tmh

	* redhat/cvsnt.spec.in: ANOTHER redhat fix.  Close to giving up on
	  the RPM

2004-08-11 16:34  tmh

	* acinclude.m4, aclocal.m4, configure: More RH9 fixes.	I'm not
	  sure if this is going to work. If not I'll forget redhat sserver
	  support.

2004-08-11 16:00  tmh

	* configure, configure.in, protocols/ntlm/ntlm.h,
	  redhat/cvsnt.spec: RH9 fixes.  It's fairly broken w/kerberos and
	  openssl so we've got to munge the config file to cope with it

2004-08-11 14:46  tmh

	* ChangeLog: Update changelog

2004-08-11 01:09  tmh

	* src/mkmodules.c, src/release.c, version_no.h: Remove some merging
	  from the main branch, and put the release stuff in just in case a
	  51c ever makes it out of the door

2004-08-06 01:11  tmh

	* src/: checkout.c, cvs.h, main.c, root.c, server.c, subr.c: Some
	  chroot fixes

2004-08-05 20:58  tmh

	* src/server.c: chroot fix

2004-08-05 20:54  tmh

	* redhat/: cvsnt-pam, cvsnt-xinetd, cvsnt.spec.in: Redhat updates

2004-08-05 19:59  tmh

	* solaris/: build_solaris64, make_package: solaris build

2004-08-05 19:57  tmh

	* hpux/: build_hpux32, build_hpux64, cvsnt.psf, make_package: hpux
	  build

2004-08-04 15:36  tmh

	* doc/cvs.dbk: doc updates

2004-08-04 15:26  tmh

	* doc/cvs.dbk: Update doc

2004-08-04 13:02  tmh

	* plink/: BE_ALL.C, BE_CVSNT.C, BE_NONE.C, BE_NOSSH.C, BLURB.BUT,
	  CHARSET.H, CMDGEN.C, CMDLINE.C, CONFIG.BUT, CONFIG.C, CONSOLE.C,
	  CONSOLE_CVSNT.C, DIALOG.C, DIALOG.H, ENUM.C, ERRORS.BUT, FAQ.BUT,
	  FEEDBACK.BUT, FROMUCS.C, GS.BUT, GTKCOLS.C, GTKCOLS.H, GTKDLG.C,
	  GTKPANEL.C, GTKPANEL.H, IMPORT.C, INT64.C, INT64.H, INTERNAL.H,
	  INTRO.BUT, KH2REG.PY, LDISC.C, LDISC.H, LDISCUCS.C, LICENCE,
	  LICENCE.BUT, LOCALENC.C, LOGGING.C, MAC.C, MAC.H, MACABOUT.C,
	  MACCTRLS.C, MACDLG.C, MACENC.C, MACEVLOG.C, MACMISC.C, MACNET.C,
	  MACNOISE.C, MACPGEN.C, MACPGEN.R, MACPGKEY.C, MACPGRID.H,
	  MACRESID.H, MACSTORE.C, MACSTUFF.H, MACTERM.C, MACUCS.C,
	  MAC_RES.R, MAKEFILE, MAKEFILE.BOR, MAKEFILE.CYG, MAKEFILE.GTK,
	  MAKEFILE.LCC, MAKEFILE.MPW, MAKEFILE.VC, MIMEENC.C, MISC.C,
	  MISC.H, MKFILES.PL, MKPUTTY.MPW, MKUNXARC.SH, MTCPNET.C,
	  NETWORK.H, NOISE.C, OTNET.C, PAGEANT.BUT, PAGEANT.C, PAGEANT.DSP,
	  PAGEANT.ICO, PAGEANT.MFT, PAGEANT.RC, PAGEANTC.C, PAGEANTS.ICO,
	  PLINK.1, PLINK.BUT, PLINK.C, PLINK.RC, PLINK_CVSNT.C, PORTFWD.C,
	  PPROXY.C, PRINTING.C, PROXY.C, PROXY.H, PSCP.BUT, PSCP.DSP,
	  PSFTP.BUT, PSFTP.C, PSFTP.DSP, PSFTP.H, PTERM.1, PTERM.C,
	  PTERMM.C, PTY.C, PUBKEY.BUT, PUTTY.1, PUTTY.DSP, PUTTY.DSW,
	  PUTTY.H, PUTTY.ICO, PUTTY.ISS, PUTTY.MFT, PUTTYCFG.ICO,
	  PUTTYGEN.1, PUTTYGEN.C, PUTTYGEN.DSP, PUTTYGEN.ICO, PUTTYGEN.MFT,
	  PUTTYGEN.RC, PUTTYMEM.H, PUTTYPS.H, PUTTYTEL.1, PUTTYTEL.DSP,
	  RAW.C, README, README.MAC, README.TXT, RECIPE, RESOURCE.H,
	  RLOGIN.C, SBCS.C, SBCS.DAT, SBCSDAT.C, SBCSGEN.PL, SCP.C,
	  SCP.ICO, SCP.RC, SETTINGS.C, SFTP.C, SFTP.H, SIGNAL.C, SITE.BUT,
	  SIZETIP.C, SLOOKUP.C, SSH.C, SSH.H, SSHAES.C, SSHBLOWF.C,
	  SSHBN.C, SSHCRC.C, SSHCRCDA.C, SSHDES.C, SSHDH.C, SSHDSS.C,
	  SSHDSSG.C, SSHMD5.C, SSHPRIME.C, SSHPUBK.C, SSHRAND.C, SSHRSA.C,
	  SSHRSAG.C, SSHSH512.C, SSHSHA.C, SSHZLIB.C, STORAGE.H, STRICMP.C,
	  TELNET.C, TERMINAL.C, TERMINAL.H, TESTBACK.C, TOUCS.C, TREE234.C,
	  TREE234.H, UNICODE.C, UNIX.H, USING.BUT, UTF8.C, UXAGENTC.C,
	  UXCFG.C, UXCONS.C, UXGEN.C, UXMISC.C, UXNET.C, UXNOISE.C,
	  UXPLINK.C, UXPRINT.C, UXPROXY.C, UXPUTTY.C, UXSEL.C, UXSFTP.C,
	  UXSTORE.C, UXUCS.C, UX_X11.C, VERSION.C, VERSION.R, VSNPRINT.C,
	  WCWIDTH.C, WEBSITE.URL, WILDCARD.C, WINCFG.C, WINCTRLS.C,
	  WINDEFS.C, WINDLG.C, WINDOW.C, WINHELP.H, WINMISC.C, WINNET.C,
	  WINSFTP.C, WINSTORE.C, WINSTUFF.H, WINUTILS.C, WIN_RES.H,
	  WIN_RES.RC, X11FWD.C, XENC.C, XKEYSYM.C, be_cvsnt.c,
	  console_cvsnt.c, plink.dsp, plink.vcproj, plink_cvsnt.c,
	  putty/CHECKLST.txt, putty/LATEST.VER, putty/LICENCE,
	  putty/MODULE, putty/Makefile.bor, putty/Makefile.cyg,
	  putty/Makefile.lcc, putty/Makefile.vc, putty/README,
	  putty/README.txt, putty/Recipe, putty/be_all.c, putty/be_none.c,
	  putty/be_nossh.c, putty/cmdgen.c, putty/cmdline.c,
	  putty/config.c, putty/console.c, putty/dialog.c, putty/dialog.h,
	  putty/import.c, putty/int64.c, putty/int64.h, putty/ldisc.c,
	  putty/ldisc.h, putty/ldiscucs.c, putty/logging.c, putty/manifest,
	  putty/misc.c, putty/misc.h, putty/mkfiles.pl, putty/mksrcarc.sh,
	  putty/mkunxarc.sh, putty/network.h, putty/noise.c,
	  putty/pageant.c, putty/pageant.ico, putty/pageant.mft,
	  putty/pageant.rc, putty/pageantc.c, putty/pageants.ico,
	  putty/plink.c, putty/plink.rc, putty/portfwd.c, putty/pproxy.c,
	  putty/printing.c, putty/proxy.c, putty/proxy.h, putty/psftp.c,
	  putty/psftp.h, putty/putty.h, putty/putty.ico, putty/putty.iss,
	  putty/putty.mft, putty/puttycfg.ico, putty/puttygen.c,
	  putty/puttygen.ico, putty/puttygen.mft, putty/puttygen.rc,
	  putty/puttymem.h, putty/puttyps.h, putty/raw.c, putty/resource.h,
	  putty/rlogin.c, putty/scp.c, putty/scp.ico, putty/scp.rc,
	  putty/settings.c, putty/sftp.c, putty/sftp.h, putty/sizetip.c,
	  putty/ssh.c, putty/ssh.h, putty/sshaes.c, putty/sshblowf.c,
	  putty/sshbn.c, putty/sshcrc.c, putty/sshcrcda.c, putty/sshdes.c,

contrib/rcs2log.sh
#! /bin/sh

# RCS to ChangeLog generator

# Generate a change log prefix from RCS files (perhaps in the CVS repository)
# and the ChangeLog (if any).
# Output the new prefix to standard output.
# You can edit this prefix by hand, and then prepend it to ChangeLog.

# Ignore log entries that start with `#'.
# Clump together log entries that start with `{topic} ',
# where `topic' contains neither white space nor `}'.

Help='The default FILEs are the files registered under the working directory.
Options:

  -c CHANGELOG  Output a change log prefix to CHANGELOG (default ChangeLog).
  -h HOSTNAME  Use HOSTNAME in change log entries (default current host).
  -i INDENT  Indent change log lines by INDENT spaces (default 8).
  -l LENGTH  Try to limit log lines to LENGTH characters (default 79).
  -R  If no FILEs are given and RCS is used, recurse through working directory.
  -r OPTION  Pass OPTION to subsidiary log command.
  -t TABWIDTH  Tab stops are every TABWIDTH characters (default 8).
  -u "LOGIN<tab>FULLNAME<tab>MAILADDR"  Assume LOGIN has FULLNAME and MAILADDR.
  -v  Append RCS revision to file names in log lines.
  --help  Output help.
  --version  Output version number.

Report bugs to <bug-gnu-emacs@gnu.org>.'

Id='$Id: rcs2log.sh,v 1.6 2003/02/07 17:45:08 tmh Exp $'

# Copyright 1992, 93, 94, 95, 96, 97, 1998 Free Software Foundation, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

Copyright='Copyright 1998 Free Software Foundation, Inc.
This program comes with NO WARRANTY, to the extent permitted by law.
You may redistribute copies of this program
under the terms of the GNU General Public License.
For more information about these matters, see the files named COPYING.
Author: Paul Eggert <eggert@twinsun.com>'

tab='	'
nl='
'

# Parse options.

# defaults
: ${AWK=awk}
: ${TMPDIR=/tmp}
changelog=ChangeLog # change log file name
datearg= # rlog date option
hostname= # name of local host (if empty, will deduce it later)
indent=8 # indent of log line
length=79 # suggested max width of log line
logins= # login names for people we know fullnames and mailaddrs of
loginFullnameMailaddrs= # login<tab>fullname<tab>mailaddr triplets
logTZ= # time zone for log dates (if empty, use local time)
recursive= # t if we want recursive rlog
revision= # t if we want revision numbers
rlog_options= # options to pass to rlog
tabwidth=8 # width of horizontal tab

while :
do
	case $1 in
	-c)	changelog=${2?}; shift;;
	-i)	indent=${2?}; shift;;
	-h)	hostname=${2?}; shift;;
	-l)	length=${2?}; shift;;
	-[nu])	# -n is obsolescent; it is replaced by -u.
		case $1 in
		-n)	case ${2?}${3?}${4?} in
			*"$tab"* | *"$nl"*)
				echo >&2 "$0: -n '$2' '$3' '$4': tabs, newlines not allowed"
				exit 1
			esac
			case $loginFullnameMailaddrs in
			'') loginFullnameMailaddrs=$2$tab$3$tab$4;;
			?*) loginFullnameMailaddrs=$loginFullnameMailaddrs$nl$2$tab$3$tab$4
			esac
			shift; shift; shift;;
		-u)
			# If $2 is not tab-separated, use colon for separator.
			case ${2?} in
			*"$nl"*)
				echo >&2 "$0: -u '$2': newlines not allowed"
				exit 1;;
			*"$tab"*)
				t=$tab;;
			*)
				t=:
			esac
			case $2 in
			*"$t"*"$t"*"$t"*)
				echo >&2 "$0: -u '$2': too many fields"
				exit 1;;
			*"$t"*"$t"*)
				;;
			*)
				echo >&2 "$0: -u '$2': not enough fields"
				exit 1
			esac
			case $loginFullnameMailaddrs in
			'') loginFullnameMailaddrs=$2;;
			?*) loginFullnameMailaddrs=$loginFullnameMailaddrs$nl$2
			esac
			shift
		esac
		case $logins in
		'') logins=$login;;
		?*) logins=$logins$nl$login
		esac
		;;
	-r)
		case $rlog_options in
		'') rlog_options=${2?};;
		?*) rlog_options=$rlog_options$nl${2?}
		esac
		shift;;
	-R)	recursive=t;;
	-t)	tabwidth=${2?}; shift;;
	-v)	revision=t;;
	--version)
		set $Id
		rcs2logVersion=$3
		echo >&2 "rcs2log (GNU Emacs) $rcs2logVersion$nl$Copyright"
		exit 0;;
	-*)	echo >&2 "Usage: $0 [OPTION]... [FILE ...]$nl$Help"
		case $1 in
		--help) exit 0;;
		*) exit 1
		esac;;
	*)	break
	esac
	shift
done

month_data='
	m[0]="Jan"; m[1]="Feb"; m[2]="Mar"
	m[3]="Apr"; m[4]="May"; m[5]="Jun"
	m[6]="Jul"; m[7]="Aug"; m[8]="Sep"
	m[9]="Oct"; m[10]="Nov"; m[11]="Dec"
'


# Put rlog output into $rlogout.

# If no rlog options are given,
# log the revisions checked in since the first ChangeLog entry.
# Since ChangeLog is only by date, some of these revisions may be duplicates of
# what's already in ChangeLog; it's the user's responsibility to remove them.
case $rlog_options in
'')
	if test -s "$changelog"
	then
		e='
			/^[0-9]+-[0-9][0-9]-[0-9][0-9]/{
				# ISO 8601 date
				print $1
				exit
			}
			/^... ... [ 0-9][0-9] [ 0-9][0-9]:[0-9][0-9]:[0-9][0-9] [0-9]+ /{
				# old-fashioned date and time (Emacs 19.31 and earlier)
				'"$month_data"'
				year = $5
				for (i=0; i<=11; i++) if (m[i] == $2) break
				dd = $3
				printf "%d-%02d-%02d\n", year, i+1, dd
				exit
			}
		'
		d=`$AWK "$e" <"$changelog"` || exit
		case $d in
		?*) datearg="-d>$d"
		esac
	fi
esac

# Use TZ specified by ChangeLog local variable, if any.
if test -s "$changelog"
then
	extractTZ='
		/^.*change-log-time-zone-rule['"$tab"' ]*:['"$tab"' ]*"\([^"]*\)".*/{
			s//\1/; p; q
		}
		/^.*change-log-time-zone-rule['"$tab"' ]*:['"$tab"' ]*t.*/{
			s//UTC0/; p; q
		}
	'
	logTZ=`tail "$changelog" | sed -n "$extractTZ"`
	case $logTZ in
	?*) TZ=$logTZ; export TZ
	esac
fi

# If CVS is in use, examine its repository, not the normal RCS files.
if test ! -f CVS/Repository
then
	rlog=rlog
	repository=
else
	rlog='cvs -q log'
	repository=`sed 1q <CVS/Repository` || exit
	test ! -f CVS/Root || CVSROOT=`cat <CVS/Root` || exit
	case $CVSROOT in
	*:/*)
		# remote repository
		;;
	*)
		# local repository
		case $repository in
		/*) ;;
		*) repository=${CVSROOT?}/$repository
		esac
		if test ! -d "$repository"
		then
			echo >&2 "$0: $repository: bad repository (see CVS/Repository)"
			exit 1
		fi
	esac
fi

# Use $rlog's -zLT option, if $rlog supports it.
case `$rlog -zLT 2>&1` in
*' option'*) ;;
*)
	case $rlog_options in
	'') rlog_options=-zLT;;
	?*) rlog_options=-zLT$nl$rlog_options
	esac
esac

# With no arguments, examine all files under the RCS directory.
case $# in
0)
	case $repository in
	'')
		oldIFS=$IFS
		IFS=$nl
		case $recursive in
		t)
			RCSdirs=`find . -name RCS -type d -print`
			filesFromRCSfiles='s|,v$||; s|/RCS/|/|; s|^\./||'
			files=`
				{
					case $RCSdirs in
					?*) find $RCSdirs \
							-type f \
							! -name '*_' \
							! -name ',*,' \
							! -name '.*_' \
							! -name .rcsfreeze.log \
							! -name .rcsfreeze.ver \
							-print
					esac
					find . -name '*,v' -print
				} |
				sort -u |
				sed "$filesFromRCSfiles"
			`;;
		*)
			files=
			for file in RCS/.* RCS/* .*,v *,v
			do
				case $file in
				RCS/. | RCS/.. | RCS/,*, | RCS/*_) continue;;
				RCS/.rcsfreeze.log | RCS/.rcsfreeze.ver) continue;;
				RCS/.\* | RCS/\* | .\*,v | \*,v) test -f "$file" || continue;;
				RCS/*,v | RCS/.*,v) ;;
				RCS/* | RCS/.*) test -f "$file" || continue
				esac
				case $files in
				'') files=$file;;
				?*) files=$files$nl$file
				esac
			done
			case $files in
			'') exit 0
			esac
		esac
		set x $files
		shift
		IFS=$oldIFS
	esac
esac

llogout=$TMPDIR/rcs2log$$l
rlogout=$TMPDIR/rcs2log$$r
trap exit 1 2 13 15
trap "rm -f $llogout $rlogout; exit 1" 0

case $datearg in
?*) $rlog $rlog_options "$datearg" ${1+"$@"} >$rlogout;;
'') $rlog $rlog_options ${1+"$@"} >$rlogout
esac || exit


# Get the full name of each author the logs mention, and set initialize_fullname
# to awk code that initializes the `fullname' awk associative array.
# Warning: foreign authors (i.e. not known in the passwd file) are mishandled;
# you have to fix the resulting output by hand.

initialize_fullname=
initialize_mailaddr=

case $loginFullnameMailaddrs in
?*)
	case $loginFullnameMailaddrs in
	*\"* | *\\*)
		sed 's/["\\]/\\&/g' >$llogout <<EOF || exit
$loginFullnameMailaddrs
EOF
		loginFullnameMailaddrs=`cat $llogout`
	esac

	oldIFS=$IFS
	IFS=$nl
	for loginFullnameMailaddr in $loginFullnameMailaddrs
	do
		case $loginFullnameMailaddr in
		*"$tab"*) IFS=$tab;;
		*) IFS=:
		esac
		set x $loginFullnameMailaddr
		login=$2
		fullname=$3
		mailaddr=$4
		initialize_fullname="$initialize_fullname
			fullname[\"$login\"] = \"$fullname\""
		initialize_mailaddr="$initialize_mailaddr
			mailaddr[\"$login\"] = \"$mailaddr\""
	done
	IFS=$oldIFS
esac

case $llogout in
?*) sort -u -o $llogout <<EOF || exit
$logins
EOF
esac
output_authors='/^date: / {
	if ($2 ~ /^[0-9]*[-\/][0-9][0-9][-\/][0-9][0-9]$/ && $3 ~ /^[0-9][0-9]:[0-9][0-9]:[0-9][0-9][-+0-9:]*;$/ && $4 == "author:" && $5 ~ /^[^;]*;$/) {
		print substr($5, 1, length($5)-1)
	}
}'
authors=`
	$AWK "$output_authors" <$rlogout |
	case $llogout in
	'') sort -u;;
	?*) sort -u | comm -23 - $llogout
	esac
`
case $authors in
?*)
	cat >$llogout <<EOF || exit
$authors
EOF
	initialize_author_script='s/["\\]/\\&/g; s/.*/author[\"&\"] = 1/'
	initialize_author=`sed -e "$initialize_author_script" <$llogout`
	awkscript='
		BEGIN {
			alphabet = "abcdefghijklmnopqrstuvwxyz"
			ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
			'"$initialize_author"'
		}
		{
			if (author[$1]) {
				fullname = $5
				if (fullname ~ /[0-9]+-[^(]*\([0-9]+\)$/) {
					# Remove the junk from fullnames like "0000-Admin(0000)".
					fullname = substr(fullname, index(fullname, "-") + 1)
					fullname = substr(fullname, 1, index(fullname, "(") - 1)
				}
				if (fullname ~ /,[^ ]/) {
					# Some sites put comma-separated junk after the fullname.
					# Remove it, but leave "Bill Gates, Jr" alone.
					fullname = substr(fullname, 1, index(fullname, ",") - 1)
				}
				abbr = index(fullname, "&")
				if (abbr) {
					a = substr($1, 1, 1)
					A = a
					i = index(alphabet, a)
					if (i) A = substr(ALPHABET, i, 1)
					fullname = substr(fullname, 1, abbr-1) A substr($1, 2) substr(fullname, abbr+1)
				}

				# Quote quotes and backslashes properly in full names.
				# Do not use gsub; traditional awk lacks it.
				quoted = ""
				rest = fullname
				for (;;) {
					p = index(rest, "\\")
					q = index(rest, "\"")
					if (p) {
						if (q && q<p) p = q
					} else {
						if (!q) break
						p = q
					}
					quoted = quoted substr(rest, 1, p-1) "\\" substr(rest, p, 1)
					rest = substr(rest, p+1)
				}

				printf "fullname[\"%s\"] = \"%s%s\"\n", $1, quoted, rest
				author[$1] = 0
			}
		}
	'

	initialize_fullname=`
		{
			(getent passwd $authors) ||
			(
				cat /etc/passwd
				for author in $authors
				do NIS_PATH= nismatch $author passwd.org_dir
				done
				ypmatch $authors passwd
			)
		} 2>/dev/null |
		$AWK -F: "$awkscript"
	`$initialize_fullname
esac


# Function to print a single log line.
# We don't use awk functions, to stay compatible with old awk versions.
# `Log' is the log message (with \n replaced by \001).
# `files' contains the affected files.
printlogline='{

	# Following the GNU coding standards, rewrite
	#	* file: (function): comment
	# to
	#	* file (function): comment
	if (Log ~ /^\([^)]*\): /) {
		i = index(Log, ")")
		files = files " " substr(Log, 1, i)
		Log = substr(Log, i+3)
	}

	# If "label: comment" is too long, break the line after the ":".
	sep = " "
	if ('"$length"' <= '"$indent"' + 1 + length(files) + index(Log, SOH)) sep = "\n" indent_string

	# Print the label.
	printf "%s*%s:", indent_string, files

	# Print each line of the log, transliterating \001 to \n.
	while ((i = index(Log, SOH)) != 0) {
		logline = substr(Log, 1, i-1)
		if (logline ~ /[^'"$tab"' ]/) {
			printf "%s%s\n", sep, logline
		} else {
			print ""
		}
		sep = indent_string
		Log = substr(Log, i+1)
	}
}'

# Pattern to match the `revision' line of rlog output.
rlog_revision_pattern='^revision [0-9]+\.[0-9]+(\.[0-9]+\.[0-9]+)*(['"$tab"' ]+locked by: [^'"$tab"' $,.0-9:;@]*[^'"$tab"' $,:;@][^'"$tab"' $,.0-9:;@]*;)?['"$tab"' ]*$'

case $hostname in
'')
	hostname=`(
		hostname || uname -n || uuname -l || cat /etc/whoami
	) 2>/dev/null` || {
		echo >&2 "$0: cannot deduce hostname"
		exit 1
	}

	case $hostname in
	*.*) ;;
	*)
		domainname=`(domainname) 2>/dev/null` &&
		case $domainname in
		*.*) hostname=$hostname.$domainname
		esac
	esac
esac


# Process the rlog output, generating ChangeLog style entries.

# First, reformat the rlog output so that each line contains one log entry.
# Transliterate \n to \001 so that multiline entries fit on a single line.
# Discard irrelevant rlog output.
$AWK <$rlogout '
	BEGIN { repository = "'"$repository"'" }
	/^RCS file:/ {
		if (repository != "") {
			filename = $3
			if (substr(filename, 1, length(repository) + 1) == repository "/") {
				filename = substr(filename, length(repository) + 2)
			}
			if (filename ~ /,v$/) {
				filename = substr(filename, 1, length(filename) - 2)
			}
			if (filename ~ /(^|\/)Attic\/[^\/]*$/) {
				i = length(filename)
				while (substr(filename, i, 1) != "/") i--
				filename = substr(filename, 1, i - 6) substr(filename, i + 1)
			}
		}
		rev = "?"
	}
	/^Working file:/ { if (repository == "") filename = $3 }
	/'"$rlog_revision_pattern"'/, /^(-----------*|===========*)$/ {
		line = $0
		if (line ~ /'"$rlog_revision_pattern"'/) {
			rev = $2
			next
		}
		if (line ~ /^date: [0-9][- +\/0-9:]*;/) {
			date = $2
			if (date ~ /\//) {
				# This is a traditional RCS format date YYYY/MM/DD.
				# Replace "/"s with "-"s to get ISO format.
				newdate = ""
				while ((i = index(date, "/")) != 0) {
					newdate = newdate substr(date, 1, i-1) "-"
					date = substr(date, i+1)
				}
				date = newdate date
			}
			time = substr($3, 1, length($3) - 1)
			author = substr($5, 1, length($5)-1)
			printf "%s %s %s %s %s %c", filename, rev, date, time, author, 1
			rev = "?"
			next
		}
		if (line ~ /^branches: /) { next }
		if (line ~ /^(-----------*|===========*)$/) { print ""; next }
		if (line == "Initial revision" || line ~ /^file .+ was initially added on branch .+\.$/) {
			line = "New file."
		}
		printf "%s%c", line, 1
	}
' |

# Now each line is of the form
# FILENAME REVISION YYYY-MM-DD HH:MM:SS[+-TIMEZONE] AUTHOR \001LOG
#	where \001 stands for a carriage return,
#	and each line of the log is terminated by \001 instead of \n.
# Sort the log entries, first by date+time (in reverse order),
# then by author, then by log entry, and finally by file name and revision
# (just in case).
sort +2 -4r +4 +0 |

# Finally, reformat the sorted log entries.
$AWK '
	BEGIN {
		logTZ = "'"$logTZ"'"
		revision = "'"$revision"'"

		# Some awk variants do not understand "\001", so we have to
		# put the char directly in the file.
		SOH="" # <-- There is a single SOH (octal code 001) here.

		# Initialize the fullname and mailaddr associative arrays.
		'"$initialize_fullname"'
		'"$initialize_mailaddr"'

		# Initialize indent string.
		indent_string = ""
		i = '"$indent"'
		if (0 < '"$tabwidth"')
			for (;  '"$tabwidth"' <= i;  i -= '"$tabwidth"')
				indent_string = indent_string "\t"
		while (1 <= i--)
			indent_string = indent_string " "
	}

	{
		newlog = substr($0, 1 + index($0, SOH))

		# Ignore log entries prefixed by "#".
		if (newlog ~ /^#/) { next }

		if (Log != newlog || date != $3 || author != $5) {

			# The previous log and this log differ.

			# Print the old log.
			if (date != "") '"$printlogline"'

			# Logs that begin with "{clumpname} " should be grouped together,
			# and the clumpname should be removed.
			# Extract the new clumpname from the log header,
			# and use it to decide whether to output a blank line.
			newclumpname = ""
			sep = "\n"
			if (date == "") sep = ""
			if (newlog ~ /^\{[^'"$tab"' }]*}['"$tab"' ]/) {
				i = index(newlog, "}")
				newclumpname = substr(newlog, 1, i)
				while (substr(newlog, i+1) ~ /^['"$tab"' ]/) i++
				newlog = substr(newlog, i+1)
				if (clumpname == newclumpname) sep = ""
			}
			printf sep
			clumpname = newclumpname

			# Get ready for the next log.
			Log = newlog
			if (files != "")
				for (i in filesknown)
					filesknown[i] = 0
			files = ""
		}
		if (date != $3  ||  author != $5) {
			# The previous date+author and this date+author differ.
			# Print the new one.
			date = $3
			time = $4
			author = $5

			zone = ""
			if (logTZ && ((i = index(time, "-")) || (i = index(time, "+"))))
				zone = " " substr(time, i)

			# Print "date[ timezone]  fullname  <email address>".
			# Get fullname and email address from associative arrays;
			# default to author and author@hostname if not in arrays.
			if (fullname[author])
				auth = fullname[author]
			else
				auth = author
			printf "%s%s  %s  ", date, zone, auth
			if (mailaddr[author])
				printf "<%s>\n\n", mailaddr[author]
			else
				printf "<%s@%s>\n\n", author, "'"$hostname"'"
		}
		if (! filesknown[$1]) {
			filesknown[$1] = 1
			if (files == "") files = " " $1
			else files = files ", " $1
			if (revision && $2 != "?") files = files " " $2
		}
	}
	END {
		# Print the last log.
		if (date != "") {
			'"$printlogline"'
			printf "\n"
		}
	}
' &&


# Exit successfully.

exec rm -f $llogout $rlogout

# Local Variables:
# tab-width:4
# End:
